src/storage/txn/scheduler.rs

```diff
@@ -154,109 +153,154 @@ pub struct Scheduler<E: Engine> {
     high_priority_pool: SchedPool<E>,
 
     // used to control write flow
-    running_write_bytes: usize,
+    running_write_bytes: AtomicUsize,
+}
+
+#[inline]
+fn id_index(cid: u64) -> usize {
+    cid as usize % TASKS_SLOTS_NUM
+}
+
+impl<E: Engine> SchedulerInner<E> {
+    /// Generates the next command ID.
+    #[inline]
+    fn gen_id(&self) -> u64 {
+        let id = self.id_alloc.fetch_add(1, Ordering::Relaxed);
+        id + 1
+    }
+
+    fn dequeue_task(&self, cid: u64) -> Task {
+        let mut tasks = self.task_contexts[id_index(cid)].lock().unwrap();
+        let task = tasks.get_mut(&cid).unwrap().task.take().unwrap();
```

Comment:
I think use `remove` directly is better.

Reply:
There is an `Option<Task>` in `TaskContext`. When getting a snapshot we need to get the inner `Task` but leave the `TaskContext`.